go/updater/trainingJobUpdater.go

```diff
@@ -48,34 +53,417 @@ type TrainingJobUpdater struct {
+}
+
+// GetStatus get TrainingJob status from trainers.
+func (updater *TrainingJobUpdater) GetStatus() (*padv1.TrainingJobStatus, error) {
+
+	status := updater.status
+
+	j, err := updater.kubeClient.BatchV1().Jobs(updater.job.Namespace).
+		Get(updater.job.Spec.Trainer.ReplicaSpec.Name, v1.GetOptions{})
+	if err != nil {
+		log.Error("get trainer error:", err.Error())
+		return &status, err
+	}
+
+	status.ReplicaStatuses, err = updater.getTrainerReplicaStatuses()
 	if err != nil {
-		return nil, err
+		log.Error("get trainer replica status error:", err.Error())
+	}
+
+	if updater.job.Spec.FaultTolerant {
+		// TODO(ZhengQi): should to confirm when job done
+		if j.Status.Failed == *updater.job.Spec.Trainer.ReplicaSpec.Spec.Parallelism {
+			status.Phase = padv1.TrainingJobPhaseFailed
+			status.Reason = "all trainer have failed!"
+		} else {
+			if j.Status.Succeeded != 0 && j.Status.Active == 0 {
+				status.Phase = padv1.TrainingJobPhaseSucceeded
+				status.Reason = "Success!"
+			}
+		}
+	} else {
+		if j.Status.Failed != 0 {
+			status.Phase = padv1.TrainingJobPhaseFailed
+			status.Reason = "at least one trainer failed!"
+		} else {
+			if j.Status.Succeeded == *updater.job.Spec.Trainer.ReplicaSpec.Spec.Parallelism && j.Status.Active == 0 {
+				status.Phase = padv1.TrainingJobPhaseSucceeded
+				status.Reason = "all trainer have succeeded!"
+			}
+		}
+	}
+	return &status, nil
+}
+
+// Convert is main process to convert TrainingJob to desire status.
+func (updater *TrainingJobUpdater) Convert(ticker *time.Ticker) {
+	log.Infof("convert status, namespace=%v name=%v: ", updater.job.Namespace, updater.job.Name)
+
+	if updater.status.Phase == padv1.TrainingJobPhaseRunning {
+		status, err := updater.GetStatus()
+		if err != nil {
+			log.Error("get current status of trainer from k8s error:", err.Error())
+			return
+		}
+		updater.status = *status.DeepCopy()
+		log.Infof("Current status namespace=%v name=%v status=%v : ", updater.job.Namespace, updater.job.Name, status)
+		err = updater.updateCRDStatus()
+		if err != nil {
+			log.Warning("get current status to update trainingJob status error: ", err.Error())
+		}
+		if updater.status.Phase == padv1.TrainingJobPhaseSucceeded || updater.status.Phase == padv1.TrainingJobPhaseFailed {
+			log.Infof("Release Resource namespace=%v name=%v: ", updater.job.Namespace, updater.job.Name)
+			if updater.job.Spec.FaultTolerant {
+				log.Infof("Release master, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+				if err := updater.releaseMaster(); err != nil {
+					log.Error(err.Error())
+				}
+			}
+			log.Infof("Release pserver, namespace=%v name=%v", updater.job.Namespace, updater.job.Spec.Trainer.ReplicaSpec.Name)
+			if err := updater.releasePserver(); err != nil {
+				log.Error(err.Error())
+			}
+		}
+	}
+
+	if updater.status.Phase == padv1.TrainingJobPhaseSucceeded || updater.status.Phase == padv1.TrainingJobPhaseFailed {
+		if ticker != nil {
```

Comment:
Maybe move this stop ticker logic to line 466, `Convert` function does not have much to do with `ticker`.

Reply:
Done. Thank you very much for your comments, and I have fixed all of the comments you mentioned above.